perf_test/sparse/KokkosSparse_gs.cpp

```diff
@@ -187,92 +188,174 @@ void runGS(const GS_Parameters& params) {
+  numericComputeTimeTotal = timer.seconds();
 
-  double applyLaunchTime = timer.seconds();
+  /////////////////// Apply /////////////////
+  timer.reset();
+  for (int i = 0; i < params.nstreams; i++) {
+    auto blk_A     = DiagBlks[i];
+    auto blk_nrows = blk_A.numRows();
+    // Last two parameters are damping factor (should be 1) and sweeps
+    switch (params.direction) {
+      case GS_SYMMETRIC:
+        KokkosSparse::Experimental::symmetric_gauss_seidel_apply(
+            instances[i], &kh[i], blk_nrows, blk_nrows, blk_A.graph.row_map,
+            blk_A.graph.entries, blk_A.values, x[i], b[i], true, true, 1.0,
+            params.sweeps);
+        break;
+      case GS_FORWARD:
+        KokkosSparse::Experimental::forward_sweep_gauss_seidel_apply(
+            instances[i], &kh[i], blk_nrows, blk_nrows, blk_A.graph.row_map,
+            blk_A.graph.entries, blk_A.values, x[i], b[i], true, true, 1.0,
+            params.sweeps);
+        break;
+      case GS_BACKWARD:
+        KokkosSparse::Experimental::backward_sweep_gauss_seidel_apply(
+            instances[i], &kh[i], blk_nrows, blk_nrows, blk_A.graph.row_map,
+            blk_A.graph.entries, blk_A.values, x[i], b[i], true, true, 1.0,
+            params.sweeps);
+        break;
+    }
+  }
+  applyLaunchTimeTotal = timer.seconds();
   timer.reset();
   Kokkos::fence();
-  double applyComputeTime = timer.seconds();
+  applyComputeTimeTotal = timer.seconds();
   timer.reset();
-  kh.destroy_gs_handle();
-  // Now, compute the 2-norm of residual
-  scalar_view_t res("Ax-b", nrows);
-  Kokkos::deep_copy(res, b);
-  scalar_t alpha = Kokkos::reduction_identity<scalar_t>::prod();
-  scalar_t beta  = -alpha;
-  KokkosSparse::spmv<scalar_t, crsMat_t, scalar_view_t, scalar_t,
-                     scalar_view_t>("N", alpha, A, x, beta, res);
-  double resnorm = KokkosBlas::nrm2(res);
-  std::cout << "\n*** Symbolic launch time: " << symbolicLaunchTime << '\n';
-  std::cout << "\n*** Symbolic compute time: " << symbolicComputeTime << '\n';
-  std::cout << "\n*** Numeric launch time: " << numericLaunchTime << '\n';
-  std::cout << "\n*** Numeric compute time: " << numericComputeTime << '\n';
-  std::cout << "\n*** Apply launch time: " << applyLaunchTime << '\n';
-  std::cout << "\n*** Apply compute time: " << applyComputeTime << '\n';
-  // note: this still works if the solution diverges
-  std::cout << "Relative res norm: " << resnorm / bnorm << '\n';
+
+  for (int i = 0; i < params.nstreams; i++) {
+    auto blk_A     = DiagBlks[i];
+    auto blk_nrows = blk_A.numRows();
+    kh[i].destroy_gs_handle();
+    // Now, compute the 2-norm of residual
+    scalar_view_t res("Ax-b", blk_nrows);
+    Kokkos::deep_copy(instances[i], res, b[i]);
+    double bnorm   = KokkosBlas::nrm2(instances[i], b[i]);
+    scalar_t alpha = Kokkos::reduction_identity<scalar_t>::prod();
+    scalar_t beta  = -alpha;
+    KokkosSparse::spmv<exec_space, scalar_t, crsMat_t, scalar_view_t, scalar_t,
```

Comment:
Not a big deal, but the spmv template arguments should all be inferrable from the arguments (if that doesn't work, I would like to know so I can fix it :smile: ) Plus it looks a little cleaner.

Reply:
OK, exec space was not inferrable from the arguments.

Reply:
Got it, just opened #1964 for that.